File: code/535-Encode and Decode TinyURL/535-Encode and Decode TinyURL.c
// Note: This is a companion problem to the System Design problem: Design TinyURL.
// TinyURL is a URL shortening service where you enter a URL such as https://leetcode.com/problems/design-tinyurl and it returns a short URL such as http://tinyurl.com/4e9iAk. Design a class to encode a URL and decode a tiny URL.

// There is no restriction on how your encode/decode algorithm should work. You just need to ensure that a URL can be encoded to a tiny URL and the tiny URL can be decoded to the original URL.

// Implement the Solution class:

// Solution() Initializes the object of the system.
// String encode(String longUrl) Returns a tiny URL for the given longUrl.
// String decode(String shortUrl) Returns the original long URL for the given shortUrl. It is guaranteed that the given shortUrl was encoded by the same object.
 

// Example 1:

// Input: url = "https://leetcode.com/problems/design-tinyurl"
// Output: "https://leetcode.com/problems/design-tinyurl"

// Explanation:
// Solution obj = new Solution();
// string tiny = obj.encode(url); // returns the encoded tiny url.
// string ans = obj.decode(tiny); // returns the original url after deconding it.
 

// Constraints:

// 1 <= url.length <= 10^4
// url is guranteed to be a valid URL.

/** Encodes a URL to a shortened URL. */
typedef struct {
    char *tiny;
    char *original;
    UT_hash_handle hh;
} Hash;

Hash *hash_table = NULL;

char *s = "1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

bool find(char *url) {
    Hash *p;
    HASH_FIND(hh, hash_table, url, strlen(url), p);
    return p;
} 


void getRandomURL(char *tiny) {
    for (int i=0; i<6; ++i) {
        tiny[i] = s[rand() % 62];
    }
}
/** Encodes a URL to a shortened URL. */
char* encode(char* longUrl) {
    char *tiny = calloc(7, sizeof(char));
    do {
        getRandomURL(tiny);
    } while (find(tiny));
    Hash *p = malloc(sizeof(Hash));
    p->tiny = tiny;
    p->original = longUrl;
    HASH_ADD_STR(hash_table, tiny, p);
    return tiny;
}

/** Decodes a shortened URL to its original URL. */
char* decode(char* shortUrl) {
    Hash *p;
    HASH_FIND(hh, hash_table, shortUrl, strlen(shortUrl), p);
    return p->original;   
}

// Your functions will be called as such:
// char* s = encode(s);
// decode(s);

File: code/535-Encode and Decode TinyURL/535-Encode and Decode TinyURL.py
# -*- coding: utf-8 -*
"""
LeetCode 535 - Encode and Decode TinyURL

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
r"""
@author: Kai Chen
@file: 535-Encode and Decode TinyURL.py
@time: 2020/9/9 9:46
@desc:
Note: This is a companion problem to the System Design problem: Design TinyURL.
TinyURL is a URL shortening service where you enter a URL
such as https://leetcode.com/problems/design-tinyurl
and it returns a short URL such as http://tinyurl.com/4e9iAk.

Design the encode and decode methods for the TinyURL service.
There is no restriction on how your encode/decode algorithm should work.
You just need to ensure that a URL can be encoded to a tiny URL and the tiny URL can be decoded to the original URL.

Accepted
"""

import random
class Codec:
    codeDic = {}
    i = 0
    def encode(self, longUrl):
        """Encodes a URL to a shortened URL.

        :type longUrl: str
        :rtype: str
        """
        while self.i in self.codeDic:
            self.i = random.randint(0, 10000)
        self.codeDic[self.i] = longUrl
        return "http://tinyurl.com/" + str(self.i)

    def decode(self, shortUrl):
        """Decodes a shortened URL to its original URL.

        :type shortUrl: str
        :rtype: str
        """
        return self.codeDic.get(int(shortUrl.replace("http://tinyurl.com/", "")))

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(url))


if __name__ == '__main__':
    # print(int("http://tinyurl.com/2836".replace("http://tinyurl.com/", "")))
    s = Codec()
    short = s.encode("https://leetcode.com/problems/design-tinyurl")
    print(short)
    print(s.codeDic)
    print(s.decode(short))
